File: solutions/278.first-bad-version/first-bad-version.js
/**
 * Definition for isBadVersion()
 * 
 * @param {integer} version number
 * @return {boolean} whether the version is bad
 * isBadVersion = function(version) {
 *     ...
 * };
 */

/**
 * @param {function} isBadVersion()
 * @return {function}
 */
const solution = isBadVersion => n => {
  let head = 1
  let tail = n
  let mid
  while (tail > head + 1) {
    mid = parseInt((head + tail) / 2)
    if (isBadVersion(mid)) tail = mid
    else head = mid
  }
  return isBadVersion(head) ? head : tail
}

File: solutions/278.first-bad-version/first-bad-version.py
# The isBadVersion API is already defined for you.
# @param version, an integer
# @return a bool
# def isBadVersion(version):


class Solution:
    def firstBadVersion(self, n):
        """
        :type n: int
        :rtype: int
            *
        0 0 1 1
        1 2 3 4
        0 1 2 3
        l m   r
        """

        l = 0
        r = n - 1

        while l <= r:
            m = int((l + r)/2)

            if not isBadVersion(m):
                # good version
                l = m + 1
            else:
                # bad version
                r = m - 1

        return l
